plugins/bazaar-backend/src/service/router.ts

```diff
@@ -0,0 +1,189 @@
+/*
+ * Copyright 2020 The Backstage Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {
+  errorHandler,
+  PluginDatabaseManager,
+  resolvePackagePath,
+  useHotMemoize,
+} from '@backstage/backend-common';
+import express from 'express';
+import Router from 'express-promise-router';
+import { Logger } from 'winston';
+import Knex from 'knex';
+import { Config } from '@backstage/config';
+
+export interface RouterOptions {
+  logger: Logger;
+  database?: PluginDatabaseManager;
+  config: Config;
+}
+
+export async function createRouter(
+  options: RouterOptions,
+): Promise<express.Router> {
+  const { logger, config } = options;
+  const db = await options.database?.getClient();
+
+  const connection = config
+    .getConfig('backend')
+    .getConfig('database')
+    .getConfig('connection');
+
+  logger.info('Initializing Bazaar backend');
+
+  const database = useHotMemoize(module, () => {
+    const knex = Knex({
```

Comment:
Think you should take a look at how the `catalog-backend` or `scaffolder-backend` plugins do all this knex work. The database connection should be passed in for you so you shouldn't have to do any of this work.

Reply:
You don't need to manage any of this database part, it's provided for you and you can pass to the `createRouter` function. See the below the scaffolder does it.

https://github.com/backstage/backstage/blob/master/packages/backend/src/plugins/scaffolder.ts#L40

https://github.com/backstage/backstage/blob/master/plugins/scaffolder-backend/src/service/router.ts#L82

https://github.com/backstage/backstage/blob/9a180af86e0e694a7ed4f122d2c27a928fbc306e/plugins/scaffolder-backend/src/scaffolder/tasks/DatabaseTaskStore.ts#L59-L61

Reply:
Thank you, we've made some changes now. Is this what you meant? @benjdlambert